### PR TITLE
[dapp-connector] Extend Cardano API with more methods

### PR DESCRIPTION
**Pull request by neuodev**
Screen Shot 2021-08-12 at 1 01 35 PM

I built the `example-cardano using  `bootstrap 4`   and each button handle on method in the  `CardanoAPI` 
When click on a button will get the response and display it on the screen

1. Added `get_balance` example 
2. New functions include `get_unused_addresses / get_used_addresses/ get_change_address / get_utxos

### Example on the `get_utxos` response
Screen Shot 2021-08-12 at 1 07 22 PM
` 

### Please take a look on my commits as we have some small issues.

### CODE REVIEW

File: packages/yoroi-extension/chrome/extension/ergo-connector/api.js [hunk 1]
```diff
@@ -191,6 +259,13 @@ export async function connectorGetChangeAddress(wallet: PublicDeriver<>): Promis
   if (change !== undefined) {
     const hash = change.addr.Hash;
     await RustModule.load();
+    // Note: SimgaRust only works for ergo
+    // RustModule.walletV2 works for cardano but doesn't not have from_bytes and to_base58 methods
+    const walletType = wallet.parent.defaultToken.Metadata.type
+
+    if(walletType === 'Cardano') {
+      return hash
+    }
```
Comment (neuodev) [inline packages/yoroi-extension/chrome/extension/ergo-connector/api.js:256]: /check

File: packages/yoroi-extension/chrome/extension/ergo-connector/api.js [hunk 2]
```diff
@@ -147,31 +154,92 @@ export async function connectorGetUtxos(
...
+
+    if(walletType === 'Cardano') {
+      // Cardano does not have a function to convert from bytes to_base58
+      addresses = addresses.map(a => a.address)
+    } else {
+      addresses =  addresses.map(a => RustModule.SigmaRust.NetworkAddress
         .from_bytes(Buffer.from(a.address, 'hex'))
         .to_base58());
+    }
+
```
Comment (neuodev) [inline packages/yoroi-extension/chrome/extension/ergo-connector/api.js]: /check

File: packages/yoroi-extension/app/ergo-connector/stores/ConnectorStore.js [hunk 3]
```diff
@@ -232,23 +232,24 @@ export default class ConnectorStore extends Store<StoresMap, ActionsMap> {
...
-      const ergoWallets = wallets.filter(
-        wallet => isErgo(wallet.getParent().getNetworkInfo())
-      );
+      const filteredWallets = wallets
+      // list all wallet for now until we find a way to
+      // know which type of wallet to filter ( ergo || cardano )
+      // .filter(wallet => !isErgo(wallet.getParent().getNetworkInfo()));
 
-      await this._getTxAssets(ergoWallets);
+      await this._getTxAssets(filteredWallets);
```
Comment (neuodev) [inline packages/yoroi-extension/app/ergo-connector/stores/ConnectorStore.js]: First, when you request a connection to Yoroi it displays  by default the ergo wallets 
Now we want it to display Ergo or Cardano  wallets according to the connection request

File: packages/yoroi-extension/app/api/ada/transactions/utils.js [hunk 4]
```diff
@@ -372,6 +372,8 @@ export function asAddressedUtxo(
       }
     );
 
+    console.log(tokenTypes)
```
Comment (yushih) [inline packages/yoroi-extension/app/api/ada/transactions/utils.js]: This is not meant to go into the final PR?

File: packages/yoroi-extension/app/api/ergo/lib/transactions/utils.js [hunk 5]
```diff
@@ -109,6 +109,7 @@ export function asAddressedUtxo(
   };
 }
 
+
```
Comment (yushih) [inline packages/yoroi-extension/app/api/ergo/lib/transactions/utils.js]: Only an empty line is added.

**Review by yushih — CHANGES_REQUESTED** [review]
We can delete `packages/yoroi-ergo-connector/example/package-lock.json` ?

And please end statements with semicolons for consistency.

File: packages/yoroi-ergo-connector/example-cardano/index.html [hunk 6]
```diff
@@ -3,11 +3,40 @@
...
+          <button id="get-utxos" class="btn btn-light w-100" >Get Utxos</button>
+        </div>
+      </div>
+      <div class="w-100 d-flex justify-content-center my-2">
+        <div class="d-none" id="spinner" role="status"></div>
+      </div>
+      <div class="d-none" id="alert" role="alert"></div>
+    </div>
     <noscript>This page contains webassembly and javascript content, please enable javascript in your browser.</noscript>
+    <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.1.0/dist/js/bootstrap.bundle.min.js" integrity="sha384-U1DAWAznBHeqEIlVSCgzq+c9gqGAJn5c/t99JyeKa9xxaYpSvHU5awsuZVVFIhvj" crossorigin="anonymous"></script>
```
Comment (yushih) [inline packages/yoroi-ergo-connector/example-cardano/index.html:39]: Question: in `index.js` there is no invocation of `bootstrap` functions, so why do we need to load `bootstrap.js`?

File: packages/yoroi-extension/app/api/ada/transactions/utils.js [hunk 7]
```diff
@@ -372,6 +372,7 @@ export function asAddressedUtxo(
       }
     );
 
+
```
Comment (yushih) [inline packages/yoroi-extension/app/api/ada/transactions/utils.js:375]: Please avoid changes that only add or delete a blank line

**Review by yushih — APPROVED** [review]

File: packages/yoroi-ergo-connector/example-cardano/index.html (hunk 6 above)
Comment (neuodev) [reply packages/yoroi-ergo-connector/example-cardano/index.html:39]: I just a part of the basic setups

**Review by yushih — APPROVED** [review]

File: packages/yoroi-extension/app/ergo-connector/stores/ConnectorStore.js [hunk 8]
```diff
@@ -232,23 +232,24 @@ export default class ConnectorStore extends Store<StoresMap, ActionsMap> {
     try {
       const wallets = await getWallets({ db: persistentDb });
 
-      const ergoWallets = wallets.filter(
-        wallet => isErgo(wallet.getParent().getNetworkInfo())
-      );
+      const filteredWallets = wallets
+      // list all wallet for now until we find a way to
+      // know which type of wallet to filter ( ergo || cardano )
+      // .filter(wallet => !isErgo(wallet.getParent().getNetworkInfo()));
```
Comment (zuzunker) [inline packages/yoroi-extension/app/ergo-connector/stores/ConnectorStore.js]: @AhmedIbrahim336 , let's plz rework this part here. In the `inject.js` in the connector, depending on which function have been called: `ergo_request_read_access` or `cardano_request_read_access` - different messages are sent: `"connector_connect_request/ergo"` or `"connector_connect_request/cardano"`.

Let's change these messages to instead include the `protocol` field, like:
```
window.postMessage({
    type: "connector_connect_request",
    protocol: "ergo",
}, location.origin);```

or 

```
window.postMessage({
    type: "connector_connect_request",
    protocol: "cardano",
}, location.origin);```

And then depending on that protocol value - filter only ergo or cardano wallets please

**Review by zuzunker — CHANGES_REQUESTED** [review]

File: packages/yoroi-extension/app/ergo-connector/stores/ConnectorStore.js (hunk 8 above)
Comment (neuodev) [reply packages/yoroi-extension/app/ergo-connector/stores/ConnectorStore.js]: In progress

**Review by zuzunker — APPROVED** [review]

**Pull request merged**